### PR TITLE
ENG-371: rename mappings to triggers; single env manifest

### PR DESCRIPTION
**Pull request by itayd**
This simplifies the manifest to support a single environment, with more sophisticated environment setups to follow.

- No more mappings. Now triggers, each trigger supports a single event. Does not map connection names for now. (this was previously reviewed by @daabr , with nits. Nits fixed here)
- Support single env in manifest.
- Simplify gRPC apply service - need to be dead simple for now. Will modify as needed.
- Killed obsolete backend/apply, now manifest is in gosdk/manifest.
- CLI to support plan/execute, combined apply and validate.

This is a "combined" PR for triggers and manifest since I don't want people to have to change their manifests twice when we merge.

New example manifest:
```
version: v1

project:
  name: test1
  vars:
    - name: MEOW
      value: "meow"
  connections:
    - name: meow_http
      token: "meow"
      integration: "http"
    - name: woof_slack
      token: "woof"
      integration: "slack"
  triggers:
    - type: post
      connection: meow_http
      entrypoint: main.star:on_http_post
    - type: message
      connection: woof_slack
      entrypoint: main.star:on_slack_message
```

### CODE REVIEW

File: backend/internal/applygrpcsvc/svc.go [hunk 1]
```diff
@@ -33,25 +35,25 @@ func (s *server) Apply(ctx context.Context, req *connect.Request[applyv1.ApplyRe
...
+		logs = append(logs, fmt.Sprintf("[plan] %s", msg))
+	}))
+	if err != nil {
+		return nil, connect.NewError(connect.CodeUnknown, err)
 	}
 
-	logs, err := s.apply.Plan(ctx, msg.Manifest)
-	if err != nil {
-		return nil, connect.NewError(connect.CodeInvalidArgument, err)
+	if err = sdkmanifest.Execute(ctx, actions, s.client, func(msg string) {
```
Comment (efiShtain) [inline backend/internal/applygrpcsvc/svc.go]: extremely unlikely, 
but when you split plan and execute, you probably have to lock resources or handle race conditions

probably was here before, but need to at least think if we want to handle it later in another ticket

File: backend/internal/db/dbgorm/triggers.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+	}, nil
+}
+
+func (db *gormdb) CreateTrigger(ctx context.Context, trigger sdktypes.Trigger) error {
+	t, err := db.triggerToRecord(ctx, trigger)
+	if err != nil {
+		return err
+	}
+
+	// Assuming if eventID already exists, nothing will happen and no error
```
Comment (efiShtain) [inline backend/internal/db/dbgorm/triggers.go]:
```suggestion

```

File: backend/internal/db/dbgorm/triggers.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+	if err := db.db.WithContext(ctx).Clauses(clause.OnConflict{DoNothing: true}).Create(t).Error; err != nil {
```
Comment (efiShtain) [inline backend/internal/db/dbgorm/triggers.go]:
```suggestion
	if err := db.db.WithContext(ctx).Create(t).Error; err != nil {
```

we want to fail if you try to create a trigger with the same id again

updates should work

File: backend/internal/db/dbgorm/triggers.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+		return translateError(err)
+	}
+	return nil
+}
+
+func (db *gormdb) UpdateTrigger(ctx context.Context, trigger sdktypes.Trigger) error {
+	t, err := db.triggerToRecord(ctx, trigger)
```
Comment (efiShtain) [inline backend/internal/db/dbgorm/triggers.go]: this method allows updating all fields in a trigger

do we really want to support changing the connectionID or projectID of a trigger ? 

EventType / CodeLocation sounds valid, but once you create a trigger it is bound to a project and connection, If you want to change those, I think a better way is deleting the old one and create a new one

(maybe connection is debatable)

File: backend/internal/db/dbgorm/triggers.go [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+	if err := db.db.WithContext(ctx).Updates(t).Error; err != nil {
+		return translateError(err)
+	}
+	return nil
+}
+
+func (db *gormdb) GetTrigger(ctx context.Context, mid sdktypes.TriggerID) (sdktypes.Trigger, error) {
+	return get(db.db, ctx, scheme.ParseTrigger, "trigger_id = ?", mid.String())
+}
```
Comment (efiShtain) [inline backend/internal/db/dbgorm/triggers.go]:
```suggestion
func (db *gormdb) GetTrigger(ctx context.Context, id sdktypes.TriggerID) (sdktypes.Trigger, error) {
	return get(db.db, ctx, scheme.ParseTrigger, "trigger_id = ?", id.String())
}
```

File: backend/internal/db/dbgorm/triggers.go [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+func (db *gormdb) DeleteTrigger(ctx context.Context, mid sdktypes.TriggerID) error {
+	var m scheme.Trigger
+	if err := db.db.WithContext(ctx).Where("trigger_id = ?", mid.String()).Delete(&m).Error; err != nil {
+		return translateError(err)
+	}
+
+	return nil
+}
```
Comment (efiShtain) [inline backend/internal/db/dbgorm/triggers.go]:
```suggestion
func (db *gormdb) DeleteTrigger(ctx context.Context, id sdktypes.TriggerID) error {
	var m scheme.Trigger
	if err := db.db.WithContext(ctx).Where("trigger_id = ?", id.String()).Delete(&m).Error; err != nil {
		return translateError(err)
	}

	return nil
}
```

do we want to handle not found somehow? currently I'm not sure if delete returns an error or not if record is not found

File: backend/internal/dispatcher/eventsworkflow.go [hunk 7]
```diff
@@ -54,78 +54,74 @@ func (d *dispatcher) getEventSessionData(ctx context.Context, event sdktypes.Eve
...
 	eventType := sdktypes.GetEventType(event)
-	for _, m := range mappings {
-		mappingName := sdktypes.GetMappingModuleName(m)
+	for _, t := range triggers {
+		envID := sdktypes.GetTriggerEnvID(t)
+		triggerEventType := sdktypes.GetTriggerEventType(t)
 
-		for _, ep := range sdktypes.GetMappingEvents(m) {
-			envID := sdktypes.GetMappingEnvID(m)
+		z := z.With(zap.String("env_id", envID.String()), zap.String("event_type", triggerEventType))
```
Comment (efiShtain) [inline backend/internal/dispatcher/eventsworkflow.go:80]: if env_id can be nil, do we want to print it?

File: backend/internal/projectsgrpcsvc/svc.go [hunk 8]
```diff
@@ -58,6 +58,31 @@ func (s *server) Create(ctx context.Context, req *connect.Request[projectsv1.Cre
...
+	}
+
+	project, err := sdktypes.ProjectFromProto(msg.Project)
+	if err != nil {
+		return nil, connect.NewError(connect.CodeInvalidArgument, err)
+	}
+
+	if err := s.projects.Update(ctx, project); err != nil {
+		if errors.Is(err, sdkerrors.ErrAlreadyExists) {
+			return nil, connect.NewError(connect.CodeAlreadyExists, err)
```
Comment (efiShtain) [inline backend/internal/projectsgrpcsvc/svc.go]: already exists sounds weird error for updating action
probably not found is better

File: backend/internal/triggers/triggers.go [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+func New(z *zap.Logger, db db.DB) sdkservices.Triggers {
+	return &triggers{db: db, z: z}
+}
+
+func (m *triggers) Create(ctx context.Context, trigger sdktypes.Trigger) (sdktypes.TriggerID, error) {
+	if mid := sdktypes.GetTriggerID(trigger); mid != nil {
+		return nil, errors.New("trigger id already defined")
+	}
```
Comment (efiShtain) [inline backend/internal/triggers/triggers.go]:
```suggestion
	if tid := sdktypes.GetTriggerID(trigger); tid != nil {
		return nil, errors.New("trigger id already defined")
	}
```

File: backend/internal/triggersgrpcsvc/svc.go [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+	if err != nil {
+		return nil, connect.NewError(connect.CodeInvalidArgument, err)
+	}
+
+	cid, err := sdktypes.ParseConnectionID(msg.ConnectionId)
+	if err != nil {
+		return nil, connect.NewError(connect.CodeInvalidArgument, err)
+	}
+
+	filter := sdkservices.ListTriggersFilter{
```
Comment (efiShtain) [inline backend/internal/triggersgrpcsvc/svc.go:128]: do we have projectID filter ? 
I think that would be valuable to display

File: cmd/ak/cmd/triggers/list.go [hunk 11]
```diff
@@ -46,14 +46,14 @@ var listCmd = common.StandardCommand(&cobra.Command{
...
+		ms, err := triggers().List(ctx, f)
 		if err != nil {
 			return err
 		}
 
 		if len(ms) == 0 {
-			var dummy *sdktypes.Mapping
-			return common.FailIfNotFound(cmd, "mappings", dummy)
+			var dummy *sdktypes.Trigger
+			return common.FailIfNotFound(cmd, "triggers", dummy)
```
Comment (efiShtain) [inline cmd/ak/cmd/triggers/list.go]: dummy?

File: cmd/ak/cmd/triggers/get.go [hunk 12]
```diff
@@ -8,23 +8,23 @@ import (
...
-	Short:   "Get connection mapping details",
+	Use:     "get <trigger ID> [--fail]",
+	Short:   "Get connection trigger details",
 	Aliases: []string{"g"},
 	Args:    cobra.ExactArgs(1),
 
 	RunE: func(cmd *cobra.Command, args []string) error {
 		r := resolver.Resolver{Client: common.Client()}
-		m, _, err := r.MappingID(args[0])
+		m, _, err := r.TriggerID(args[0])
```
Comment (efiShtain) [inline cmd/ak/cmd/triggers/get.go]:
```suggestion
		t, _, err := r.TriggerID(args[0])
```

**Review by efiShtain — COMMENTED** [review]
some stuff to consider

main things that bothers me, is having sdkmanifest, 
why do we need this as an sdk ? 
who is going to use it in that way ?

File: backend/internal/applygrpcsvc/svc.go (hunk 1 above)
Comment (daabr) [reply backend/internal/applygrpcsvc/svc.go]: +1 but not for this PR, but probably not even soon. Just need to file a ticket for tracking.

File: sdk/sdkmanifest/actions.go [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+
+type Action interface {
+	keyer
+
+	Type() string
+
+	isAction()
+}
+
+// ---
```
Comment (daabr) [inline sdk/sdkmanifest/actions.go]: Multiple `// ---` markers in a file are a signal of bad file organization: the right way would be to use multiple files in a subdirectory, and give it a simple name that does NOT include its parents (`sdk` and `manifest`) as a prefix.

Also, all the structs can extend a single base struct for common fields and methods.

File: sdk/sdkmanifest/actions_json.go [hunk 14]
```diff
@@ -0,0 +1,68 @@
+package sdkmanifest
+
+import (
+	"encoding/json"
+	"fmt"
+)
+
+func (as Actions) MarshalJSON() ([]byte, error) {
+	type action struct {
```
Comment (daabr) [inline sdk/sdkmanifest/actions_json.go]: Type definitions do not belong inside functions. Due to the similar pattern, you could also convert this into a directory of separate but similar files.

File: sdk/sdkmanifest/exec_context.go [hunk 15]
```diff
@@ -0,0 +1,90 @@
+package sdkmanifest
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"go.autokitteh.dev/autokitteh/internal/resolver"
```
Comment (daabr) [inline sdk/sdkmanifest/exec_context.go]: Not for this PR, but we may want to consider moving the resolver to the SDK after we refactor SDK types and IDs. That would make more sense than this dependency inversion. Please file a ticket for tracking.

File: sdk/sdkmanifest/read.go [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+	)
+	if err != nil {
+		return nil, fmt.Errorf("validate: %w", err)
+	}
+
+	if !res.Valid() {
+		msg := strings.Join(kittehs.Transform(res.Errors(), func(err gojsonschema.ResultError) string {
+			return fmt.Sprintf("- %s: %s", err.Field(), err.Description())
+		}), "\n")
+		return nil, fmt.Errorf("invalid YAML semantics: %s", msg)
```
Comment (daabr) [inline sdk/sdkmanifest/read.go]:
```suggestion
		return nil, fmt.Errorf("invalid YAML semantics:\n%s", msg)
```

File: cmd/ak/cmd/manifest/apply.go [hunk 17]
```diff
@@ -2,100 +2,45 @@ package manifest
...
 func init() {
 	// Command-specific flags.
-	applyCmd.Flags().BoolVarP(&noValidate, "no-validate", "n", false, "don't validate before applying")
 	applyCmd.Flags().BoolVarP(&dryRun, "dry-run", "d", false, "only show plan, don't apply")
-}
-
-func err2str(err gojsonschema.ResultError) string {
-	return fmt.Sprintf("- %s: %s", err.Field(), err.Description())
+	applyCmd.Flags().BoolVarP(&noValidate, "no-validate", "n", false, "do not validate")
+	applyCmd.Flags().BoolVarP(&fromScratch, "from-scratch", "s", false, "assume no existing setup")
```
Comment (daabr) [inline cmd/ak/cmd/manifest/apply.go:49]: Why do we want to allow the `--from-scratch` flag? If we need to delete existing-but-obsolete stuff, it seems like a reasonable default, not an option that needs to specify.

File: cmd/ak/common/context.go [hunk 18]
```diff
@@ -2,12 +2,23 @@ package common
...
+	"os"
 	"time"
 )
 
-const (
-	timeout = 1 * time.Minute
-)
+var timeout = 1 * time.Minute
+
+func init() {
```
Comment (daabr) [inline cmd/ak/common/context.go:12]: I don't like this overly generic name which isn't used generically across the codebase, and that it's unrelated to the existing config mechnism, but I don't mind keeping it in this PR given that the `TODO(ENG-320)` comment still applies.

File: backend/internal/db/dbgorm/triggers.go (hunk 3 above)
Comment (daabr) [reply backend/internal/db/dbgorm/triggers.go]: @efiShtain not sure I agree. This is valid idempotent behavior, just like no-op updates.

File: cmd/ak/cmd/triggers/create.go [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+			err = fmt.Errorf("connection %q not found", connection)
+			return common.NewExitCodeError(common.NotFoundExitCode, err)
+		}
+
+		cl, err := sdktypes.ParseCodeLocation(loc)
+		if err != nil {
+			return fmt.Errorf("invalid entry-point location %q: %w", loc, err)
+		}
+
+		m, err := sdktypes.StrictTriggerFromProto(&sdktypes.TriggerPB{
```
Comment (daabr) [inline cmd/ak/cmd/triggers/create.go]:
```suggestion
		t, err := sdktypes.StrictTriggerFromProto(&sdktypes.TriggerPB{
```

File: cmd/ak/cmd/triggers/create.go [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+			EnvId:        eid.String(),
+			ConnectionId: cid.String(),
+			EventType:    event,
+			CodeLocation: cl.ToProto(),
+		})
+		if err != nil {
+			return fmt.Errorf("invalid trigger: %w", err)
+		}
+
+		mid, err := triggers().Create(context.Background(), m)
```
Comment (daabr) [inline cmd/ak/cmd/triggers/create.go]:
```suggestion
		tid, err := triggers().Create(context.Background(), t)
```

File: cmd/ak/cmd/triggers/delete.go [hunk 21]
```diff
@@ -10,27 +10,27 @@ import (
...
-	Short:   "Delete connection mapping",
+	Use:     "delete <trigger ID>",
+	Short:   "Delete connection trigger",
 	Aliases: []string{"del", "d"},
 	Args:    cobra.ExactArgs(1),
 
 	RunE: func(cmd *cobra.Command, args []string) error {
 		r := resolver.Resolver{Client: common.Client()}
-		m, id, err := r.MappingID(args[0])
+		m, id, err := r.TriggerID(args[0])
```
Comment (daabr) [inline cmd/ak/cmd/triggers/delete.go]:
```suggestion
		t, id, err := r.TriggerID(args[0])
```

File: cmd/ak/cmd/triggers/list.go [hunk 22]
```diff
@@ -46,14 +46,14 @@ var listCmd = common.StandardCommand(&cobra.Command{
 		ctx, cancel := common.LimitedContext()
 		defer cancel()
 
-		ms, err := mappings().List(ctx, f)
+		ms, err := triggers().List(ctx, f)
```
Comment (daabr) [inline cmd/ak/cmd/triggers/list.go]:
```suggestion
		ts, err := triggers().List(ctx, f)
```

File: cmd/ak/cmd/triggers/list.go (hunk 11 above)
Comment (daabr) [reply cmd/ak/cmd/triggers/list.go]: @efiShtain IIRC I wrote this because I needed a nil-by-definition-but-typed instance to pass to `FailIfNotFound`. I think it's a reasonable name.

File: cmd/ak/common/consume.go [hunk 23]
```diff
@@ -0,0 +1,27 @@
+package common
+
+import (
+	"fmt"
+	"io"
+	"os"
+)
+
+func Consume(args []string) (data []byte, path string, err error) {
```
Comment (daabr) [inline cmd/ak/common/consume.go:9]: Why not `*string` to avoid the unnecessary default case?

File: cmd/ak/cmd/manifest/apply.go [hunk 24]
```diff
@@ -2,100 +2,45 @@ package manifest
...
-	Logs       any `json:"logs"`
-	Operations any `json:"operations"`
-}
-
-var dryRun, noValidate bool
+var dryRun bool
 
 var applyCmd = common.StandardCommand(&cobra.Command{
-	Use:     "apply [file] [--no-validate] [--dry-run]",
+	Use:     "apply [file] [--no-validate] [--from-scratch] [--dry-run]",
```
Comment (daabr) [inline cmd/ak/cmd/manifest/apply.go:16]: You split the planning into a separate command - why? Anyway, `--dry-run` (i.e. plan without execute) has no meaning or usage now.

File: cmd/ak/cmd/manifest/exec.go [hunk 25]
```diff
@@ -0,0 +1,39 @@
...
+	"fmt"
+	"os"
+
+	"github.com/spf13/cobra"
+
+	"go.autokitteh.dev/autokitteh/cmd/ak/common"
+	"go.autokitteh.dev/autokitteh/sdk/sdkmanifest"
+)
+
+var execCmd = common.StandardCommand(&cobra.Command{
```
Comment (daabr) [inline cmd/ak/cmd/manifest/exec.go:14]: I don't understand, why have both `apply` and `exec`? Also, "execute" is an inaccurate name from this CLI command - we're not executing code in the regular sense, we're applying configurations.

File: cmd/ak/cmd/manifest/apply.go [hunk 26]
```diff
@@ -2,100 +2,45 @@ package manifest
...
-				gojsonschema.NewGoLoader(&root),
-			)
-			if err != nil {
-				return fmt.Errorf("validate: %w", err)
-			}
-			if !res.Valid() {
-				msg := strings.Join(kittehs.Transform(res.Errors(), err2str), "\n")
-				return fmt.Errorf("invalid YAML semantics: %s", msg)
-			}
+		actions, err := plan(data, path)
```
Comment (daabr) [inline cmd/ak/cmd/manifest/apply.go:27]: If this function is called only here, why is it implemented in a different file?

File: cmd/ak/cmd/manifest/plan.go [hunk 27]
```diff
@@ -0,0 +1,73 @@
...
+		return nil
+	},
+})
+
+func init() {
+	planCmd.Flags().BoolVarP(&noValidate, "no-validate", "n", false, "do not validate")
+	planCmd.Flags().BoolVarP(&fromScratch, "from-scratch", "s", false, "assume no existing setup")
+}
+
+func plan(data []byte, path string) (sdkmanifest.Actions, error) {
```
Comment (daabr) [inline cmd/ak/cmd/manifest/plan.go]: Did you mean to call this function in this file too?

File: cmd/ak/cmd/manifest/validate.go [hunk 28]
```diff
@@ -0,0 +1,28 @@
...
+import (
+	"github.com/spf13/cobra"
+
+	"go.autokitteh.dev/autokitteh/cmd/ak/common"
+	"go.autokitteh.dev/autokitteh/sdk/sdkmanifest"
+)
+
+var validateCmd = common.StandardCommand(&cobra.Command{
+	Use:     "validate [file]",
+	Short:   "validate manifest",
```
Comment (daabr) [inline cmd/ak/cmd/manifest/validate.go]:
```suggestion
	Short:   "Validate YAML manifest, from a file or stdin",
```

File: cmd/ak/cmd/manifest/plan.go [hunk 29]
```diff
@@ -0,0 +1,73 @@
...
+
+	"go.autokitteh.dev/autokitteh/cmd/ak/common"
+	"go.autokitteh.dev/autokitteh/sdk/sdkmanifest"
+)
+
+var noValidate, fromScratch bool
+
+var planCmd = common.StandardCommand(&cobra.Command{
+	Use:     "plan [file] [--no-validate] [--from-scratch]",
+	Short:   "plan manifest application",
```
Comment (daabr) [inline cmd/ak/cmd/manifest/plan.go]:
```suggestion
	Short:   "Dry-run for applying a YAML manifest, from a file or stdin",
```

**Review by daabr — COMMENTED** [review]
I agree with Efi - the manifest code doesn't belong in the SDK. It's not supposed to be reused or extended by users.

Also, I agree with the motivation to combine 2 breaking changes in manifest files, but it would still be more readable to implement them in separate (stacked) PRs, and merge them together - because they unrelated to each other, and big.

File: sdk/sdkmanifest/plan.go [hunk 30]
```diff
@@ -0,0 +1,427 @@
...
+	"fmt"
+	"os"
+
+	"go.autokitteh.dev/autokitteh/internal/kittehs"
+	"go.autokitteh.dev/autokitteh/sdk/sdklogger"
+	"go.autokitteh.dev/autokitteh/sdk/sdkservices"
+	"go.autokitteh.dev/autokitteh/sdk/sdktypes"
+)
+
+const defaultEnvName = "default"
```
Comment (daabr) [inline sdk/sdkmanifest/plan.go]: "default" is not a good default name, because we may have other defaults in the future. Maybe just "env"?

File: backend/internal/applygrpcsvc/svc.go (hunk 1 above)
Comment (itayd) [reply backend/internal/applygrpcsvc/svc.go]: having plan and execute separate is a good practice, not necessarily in order to not run separately but definitely for testing. also this gives the user a chance to go over the plan and maybe even prune/change it if needed. locking resources is complicated, not sure we want to do this?

File: backend/internal/db/dbgorm/triggers.go (hunk 3 above)
Comment (itayd) [reply backend/internal/db/dbgorm/triggers.go]: i think i added it back then when i thought we might have to retry adding with the same id. I'm going to adopt efi's proposal here and remove it - this shouldn't happen.

File: backend/internal/db/dbgorm/triggers.go (hunk 6 above)
Comment (itayd) [reply backend/internal/db/dbgorm/triggers.go]: i think it will fail silently, i am ok with it.

File: backend/internal/dispatcher/eventsworkflow.go (hunk 7 above)
Comment (itayd) [reply backend/internal/dispatcher/eventsworkflow.go:80]: it will print an empty string, i'm ok with it. no need to if else on that.

File: cmd/ak/cmd/manifest/apply.go (hunk 17 above)
Comment (itayd) [reply cmd/ak/cmd/manifest/apply.go:49]: tbh it's good for debugging, or maybe you want to send a list of commands to execute to start from scratch.

File: backend/internal/db/dbgorm/triggers.go (hunk 4 above)
Comment (itayd) [reply backend/internal/db/dbgorm/triggers.go]: yup. checking for changes now.

File: backend/internal/projectsgrpcsvc/svc.go (hunk 8 above)
Comment (itayd) [reply backend/internal/projectsgrpcsvc/svc.go]: shouldn't be there. removed.

File: backend/internal/triggersgrpcsvc/svc.go (hunk 10 above)
Comment (itayd) [reply backend/internal/triggersgrpcsvc/svc.go:128]: ENG-445 now

File: cmd/ak/cmd/triggers/list.go (hunk 11 above)
Comment (itayd) [reply cmd/ak/cmd/triggers/list.go]: removed dummy, explicitly set type

Comment (itayd) [reply cmd/ak/cmd/triggers/list.go]: actually scratch that. removed check entirely - failifnotfound should not be allowed on list results.

File: sdk/sdkmanifest/actions_json.go (hunk 14 above)
Comment (itayd) [reply sdk/sdkmanifest/actions_json.go]: i disagree about the "do not belong inside functions". this specific struct has no use outside this function.

File: sdk/sdkmanifest/exec_context.go (hunk 15 above)
Comment (itayd) [reply sdk/sdkmanifest/exec_context.go]: see PRD-25

File: cmd/ak/common/context.go (hunk 18 above)
Comment (itayd) [reply cmd/ak/common/context.go:12]: i think it still applies, let's leave it as is.

File: cmd/ak/common/consume.go (hunk 23 above)
Comment (itayd) [reply cmd/ak/common/consume.go:9]: not sure i follow, but if i do - you want to check if there is an arg prior to passing it to `Consume`? I think for now this is much more convinient.

File: cmd/ak/cmd/manifest/apply.go (hunk 24 above)
Comment (itayd) [reply cmd/ak/cmd/manifest/apply.go:16]: discussed in https://github.com/autokitteh/autokitteh/pull/6#discussion_r1496787189

File: cmd/ak/cmd/manifest/exec.go (hunk 25 above)
Comment (itayd) [reply cmd/ak/cmd/manifest/exec.go:14]: apply is for:
- simple single command to do everything
- backward compatible to what we have now

exec is executing actions, just executes all the actions one after another.

File: cmd/ak/cmd/manifest/apply.go (hunk 26 above)
Comment (itayd) [reply cmd/ak/cmd/manifest/apply.go:27]: oops - now used in plan cmd too.

File: cmd/ak/cmd/manifest/plan.go (hunk 27 above)
Comment (itayd) [reply cmd/ak/cmd/manifest/plan.go]: yes, done.

File: sdk/sdkmanifest/plan.go (hunk 30 above)
Comment (itayd) [reply sdk/sdkmanifest/plan.go]: it's a default environment name. i think it's apt.

File: sdk/sdkmanifest/actions.go (hunk 13 above)
Comment (itayd) [reply sdk/sdkmanifest/actions.go]: split into `internal/actions`, also split into multiple files based on object type. i think splitting it further is overkill.

File: backend/internal/applygrpcsvc/svc.go (hunk 1 above)
Comment (daabr) [reply backend/internal/applygrpcsvc/svc.go]: Maybe not locking between planning and applying, after all there's no guarantee that the user will run "apply" at all, but **during** applying it's a good idea to prevent a potential race between 2 users - but again, we don't even support multiple users right now.

Comment (daabr) [reply backend/internal/applygrpcsvc/svc.go]: Maybe add a "locked" property to projects that only AK can modify? (while it's creating/updating a project)

File: cmd/ak/cmd/triggers/list.go (hunk 11 above)
Comment (daabr) [reply cmd/ak/cmd/triggers/list.go]: No, you need this check because all `list` commands have a `--fail` flag (i.e. return code = 1 if list is empty). This is the standard way right now to implement it, by reusing the check which was originally written for `get` commands.

File: cmd/ak/cmd/manifest/apply.go [hunk 31]
```diff
@@ -2,100 +2,49 @@ package manifest
...
-	"strings"
 
 	"github.com/spf13/cobra"
-	"github.com/xeipuuv/gojsonschema"
-	"gopkg.in/yaml.v3"
 
-	"go.autokitteh.dev/autokitteh/backend/apply"
 	"go.autokitteh.dev/autokitteh/cmd/ak/common"
-	"go.autokitteh.dev/autokitteh/internal/kittehs"
+	imanifest "go.autokitteh.dev/autokitteh/internal/manifest"
```
Comment (daabr) [inline cmd/ak/cmd/manifest/apply.go:10]: Are you sure this alias is necessary? I think Go understands that if you use `manifest.Foo()` inside a package called `manifest` it should be looking for a `manifest` import inside this package, and not the package itself.

File: cmd/ak/cmd/manifest/exec.go [hunk 32]
```diff
@@ -0,0 +1,39 @@
...
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+
+	"github.com/spf13/cobra"
+
+	"go.autokitteh.dev/autokitteh/cmd/ak/common"
+	imanifest "go.autokitteh.dev/autokitteh/internal/manifest"
```
Comment (daabr) [inline cmd/ak/cmd/manifest/exec.go:11]: Same comment as in `cmd/ak/cmd/manifest/apply.go` - import aliases as confusing, unless strictly necessary. Same also in `schema.go` and `validate.go`. Feel free to resolve if it is necessary.

File: cmd/ak/cmd/manifest/exec.go (hunk 25 above)
Comment (daabr) [reply cmd/ak/cmd/manifest/exec.go:14]: Still sounds like `exec` is not a useful command for users in real-life.

After thinking about this some more, I think the most user-friendly thing to do is to behave like Terraform, which is a very well known behavior: https://developer.hashicorp.com/terraform/cli/run

tl;dr -
1. Keep the "plan" command
2. Remove the "exec" command
3. Plan and apply modes (with mutually-exclusive flags):

   * Normal (default)
   * `--destroy` (rename `--from-scratch`)
   * `--refresh-only`

If you don't want to do it in this PR, which is fine, please file a ticket that describes this task and/or references this specific comment thread.

File: cmd/ak/cmd/manifest/apply.go (hunk 24 above)
Comment (daabr) [reply cmd/ak/cmd/manifest/apply.go:16]: Closing this thread in favor of the other one.

File: cmd/ak/cmd/manifest/apply.go (hunk 17 above)
Comment (daabr) [reply cmd/ak/cmd/manifest/apply.go:49]: Closing this comment thread in favor of https://github.com/autokitteh/autokitteh/pull/6/files?diff=split&w=1#r1496711877

File: cmd/ak/cmd/triggers/delete.go [hunk 33]
```diff
@@ -10,27 +10,27 @@ import (
 )
 
 var deleteCmd = common.StandardCommand(&cobra.Command{
-	Use:     "delete <mapping ID>",
-	Short:   "Delete connection mapping",
+	Use:     "delete <trigger ID>",
+	Short:   "Delete connection trigger",
```
Comment (daabr) [inline cmd/ak/cmd/triggers/delete.go]: IIRC I commented in the previous PR that it's more accurate now to refer to triggers as "event triggers" rather than "connection trigger". Or just "trigger" without anything. Please fix in the `Short` fields of all other trigger commands too.

File: cmd/ak/common/consume.go (hunk 23 above)
Comment (daabr) [reply cmd/ak/common/consume.go:9]: You're right. OTOH, all the commands that call this have `Args: cobra.MaximumNArgs(1)` :-\

In case we use `Consume` in the future in a command that does accept more than 1 arg, we'll have to fix this anyway, so we might as well assume it's either 0 or 1, and use `if len(args) == 0 else` (and add a comment explaining why we allow this assumption).

But if you disagree, feel free to resolve this comment.

File: internal/manifest/internal/actions/trigger.go [hunk 34]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+func (a CreateTriggerAction) Type() string   { return "create_trigger" }
+func (a CreateTriggerAction) isAction()      {}
+func (a CreateTriggerAction) GetKey() string { return a.Key }
+
+func init() { registerActionType[CreateTriggerAction]() }
+
+// ---
+type UpdateTriggerAction struct {
```
Comment (daabr) [inline internal/manifest/internal/actions/trigger.go]:
```suggestion
// ---

type UpdateTriggerAction struct {
```

File: internal/manifest/exec_context.go [hunk 35]
```diff
@@ -0,0 +1,90 @@
...
+	p, err := c.client.Projects().GetByName(ctx, sdkName)
+	if err != nil {
+		return nil, err
+	}
+
+	pid := sdktypes.GetProjectID(p)
+
+	c.projects[name] = pid
+
+	return pid, nil
```
Comment (daabr) [inline internal/manifest/exec_context.go]: Feel free to keep this, but IMO when number of empty lines == number of code lines, you're using too many whitespaces.  :-P

File: internal/manifest/plan.go [hunk 36]
```diff
@@ -0,0 +1,432 @@
+package manifest
```
Comment (daabr) [inline internal/manifest/plan.go:1]: Even GitHub thinks this file is too big to review - it hides the diff by default :)

But I don't have a good suggestion for splitting it...

File: systest/testdata/manifest/apply.txtar [hunk 37]
```diff
@@ -37,21 +37,21 @@ ak connection get my_project/my_connection -J
 output equals file expected_connection_details.json
 return code == 0
 
-ak connection get connection:00000000000000000000000000000002 -J
+ak connection get connection:00000000000000000000000000000003 -J
```
Comment (daabr) [inline systest/testdata/manifest/apply.txtar:44]:
```suggestion
# TODO: Add more tests - either here, or in separate `.txtar` files:
# - Specifying an env name
# - This command's flags
# - Other manifest commands since PR #6

ak connection get connection:00000000000000000000000000000003 -J
```

**Review by daabr — COMMENTED** [review]
A few more comments and replies

File: backend/internal/applygrpcsvc/svc.go (hunk 1 above)
Comment (efiShtain) [reply backend/internal/applygrpcsvc/svc.go]: working with terraform, if you plan and save the plan result, 
you can apply the plan and are guaranteed that the end result is what you have in plan 
so in case something changed the state between plan and apply (which can be zero time, days or never), what you planned is what you get

File: backend/internal/db/dbgorm/triggers.go (hunk 4 above)
Comment (efiShtain) [reply backend/internal/db/dbgorm/triggers.go]: I think its weird to allow changing the project ID of a trigger

its not something a user would do, unless we have a dedicated trigger creation capability where we can attach triggers to projects

File: systest/testdata/manifest/apply.txtar (hunk 37 above)
Comment (itayd) [reply systest/testdata/manifest/apply.txtar:44]: done

File: internal/manifest/exec_context.go (hunk 35 above)
Comment (itayd) [reply internal/manifest/exec_context.go]: "fixed"

File: internal/manifest/internal/actions/trigger.go (hunk 34 above)
Comment (itayd) [reply internal/manifest/internal/actions/trigger.go]: done

File: cmd/ak/cmd/triggers/delete.go (hunk 33 above)
Comment (itayd) [reply cmd/ak/cmd/triggers/delete.go]: fixed

File: internal/manifest/internal/actions/trigger.go (hunk 34 above)
Comment (itayd) [reply internal/manifest/internal/actions/trigger.go]: fixed

File: cmd/ak/cmd/manifest/exec.go (hunk 25 above)
Comment (itayd) [reply cmd/ak/cmd/manifest/exec.go:14]: Worth thinking about it. Now in ENG-450.

**Comment by itayd** [comment]
Re https://github.com/autokitteh/autokitteh/pull/6#discussion_r1496948425: i think it's not worth the hassle right now. If we do the changes you forsee in the future we'll do it there.

File: backend/internal/applygrpcsvc/svc.go (hunk 1 above)
Comment (itayd) [reply backend/internal/applygrpcsvc/svc.go]: probably everything in terraform is idempotent?

File: backend/internal/db/dbgorm/triggers.go (hunk 4 above)
Comment (itayd) [reply backend/internal/db/dbgorm/triggers.go]: yup no longer allowing it

File: cmd/ak/cmd/triggers/list.go (hunk 11 above)
Comment (itayd) [reply cmd/ak/cmd/triggers/list.go]: i'll keep the failifnotfound for now, but let's discuss as part of ENG-451

**Review by daabr — APPROVED** [review]
